### PR TITLE
Support partition by

### PR DESCRIPTION
**Pull request by qingxinhome**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #3416

## What this PR does / why we need it:
The DDL statement supports the partition by clause. The partition types include:
RANGE Partitioning,
LIST Partitioning,
COLUMNS Partitioning,
RANGE COLUMNS partitioning,
LIST COLUMNS partitioning,
HASH Partitioning,
KEY Partitioning,

The construction of the partition sentence and the semantic check coding have been completed, but the partition information of DDL is not synchronized to the metadata table. Subsequently, the metadata table related to the partition will be added and the partition information will be displayed when show create table is supported, Issue#4749

### CODE REVIEW

**Review by daviszhen — APPROVED** [review]

**Review by zhangxu19830126 — APPROVED** [review]

**Comment by guguducken** [comment]
@matrix-meow rerun failed

File: pkg/sql/plan/utils.go [hunk 1]
```diff
@@ -655,3 +655,53 @@ func isConstant(e *plan.Expr) bool {
 		return false
 	}
 }
+
```
Comment (fengttt) [inline pkg/sql/plan/utils.go:658]: These code, if you really need it, should go into container types.

**Review by fengttt — CHANGES_REQUESTED** [review]
This change need to have a bvt test.

**Comment by guguducken** [comment]
@matrix-meow rerun failed yes

File: proto/plan.proto [hunk 2]
```diff
@@ -36,20 +36,21 @@ message Type {
 // double, use that, otherwise store a string representation.
 message Const {
 	bool isnull			= 1;
+	bool isMaxValue = 2;
```
Comment (ouyuanning) [inline proto/plan.proto:39]: I don't think isMaxValue be a field of Const is a good choice.
That will make constant fold complicated。

File: proto/plan.proto [hunk 3]
```diff
@@ -218,6 +219,43 @@ message PropertiesDef {
 	repeated Property properties 	= 1;
 }
 
+
+enum PartitionType {
```
Comment (ouyuanning) [inline proto/plan.proto:223]: if only used in PartitionInfo.   you can move these code to message PartitionInfo

File: pkg/sql/plan/utils.go (hunk 1 above)
Comment (qingxinhome) [reply pkg/sql/plan/utils.go:658]: Ok, no problem

File: pkg/sql/plan/build_ddl.go [hunk 4]
```diff
@@ -230,6 +247,25 @@ func buildCreateTable(stmt *tree.CreateTable, ctx CompilerContext) (*Plan, error
...
+	case *tree.HashType:
+		err = buildHashPartition(partitionBinder, partitionOp, tableDef)
+	case *tree.KeyType:
+		err = buildKeyPartition(partitionBinder, partitionOp, tableDef)
+	case *tree.RangeType:
+		err = buildRangePartition(partitionBinder, partitionOp, tableDef)
+	case *tree.ListType:
+		err = buildListPartitiion(partitionBinder, partitionOp, tableDef)
+	}
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (qingxinhome) [reply pkg/sql/plan/build_ddl.go:263]: haha，yes，you are right!

File: proto/plan.proto (hunk 3 above)
Comment (qingxinhome) [reply proto/plan.proto:223]: I've tried to do this, but the constant name generated by protobuffer is very lengthy and does not express the meaning

File: pkg/pb/txn/txn.pb.go [hunk 5]
```diff
@@ -526,9 +526,9 @@ func (m *CNOpResponse) GetPayload() []byte {
 // to TxnMethod.
 //
 // Request flow of TxnRequest as below:
-//  1. CN -> DN (TxnMethod.Read, TxnMethod.Write, TxnMethod.Commit, TxnMethod.Rollback)
-//  2. DN -> DN (TxnMethod.Prepare, TxnMethod.GetStatus, TxnMethod.CommitDNShard, TxnMethod.RollbackDNShard,
-//     TxnMethod.RemoveMetadata)
+// 1. CN -> DN (TxnMethod.Read, TxnMethod.Write, TxnMethod.Commit, TxnMethod.Rollback)
+// 2. DN -> DN (TxnMethod.Prepare, TxnMethod.GetStatus, TxnMethod.CommitDNShard, TxnMethod.RollbackDNShard,
+//             TxnMethod.RemoveMetadata)
```
[the remark this answers is not in the document]
Comment (qingxinhome) [reply pkg/pb/txn/txn.pb.go:531]: I don't know, the code is merged

**Review by aunjgr — APPROVED** [review]

**Comment by qingxinhome** [comment]
> This change need to have a bvt test.

ok， I will supplement the BVT test

File: pkg/sql/plan/utils.go [hunk 6]
```diff
@@ -655,3 +655,53 @@ func isConstant(e *plan.Expr) bool {
...
+	if t == types.T_int8 || t == types.T_int16 || t == types.T_int32 || t == types.T_int64 {
+		return true
+	}
+	return false
+}
+
+// if expr type is integer return true,else return false
+func IsIntegerType(expr *plan.Expr) bool {
+	t := types.T(expr.Typ.Id)
+	if t == types.T_int8 || t == types.T_int16 || t == types.T_int32 || t == types.T_int64 ||
```
[the remark this answers is not in the document]
Comment (qingxinhome) [reply pkg/sql/plan/utils.go]: OK， good idea！

File: pkg/sql/plan/build_ddl.go [hunk 7]
```diff
@@ -215,7 +219,20 @@ func buildCreateTable(stmt *tree.CreateTable, ctx CompilerContext) (*Plan, error
...
-		return nil, errors.New(errno.SQLStatementNotYetComplete, fmt.Sprintf("partition unsupport now; statement: '%v'", tree.String(stmt, dialect.MYSQL)))
+		nodeID := builder.appendNode(&plan.Node{
+			NodeType:    plan.Node_TABLE_SCAN,
+			Cost:        nil,
+			ObjRef:      nil,
+			TableDef:    createTable.TableDef,
+			BindingTags: []int32{builder.genNewTag()},
+		}, bindContext)
+
+		err = builder.addBinding(nodeID, tree.AliasClause{}, bindContext)
```
[the remark this answers is not in the document]
Comment (qingxinhome) [reply pkg/sql/plan/build_ddl.go:231]: OK， you are right!

File: proto/plan.proto (hunk 2 above)
Comment (qingxinhome) [reply proto/plan.proto]: OK, I will add a new expression to handle ‘MAXVALUE’

File: test/cases/ddl/table_partition.test [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+	a INT NULL,
+	b INT NULL
+)
+PARTITION BY LIST COLUMNS(a,b) (
+	PARTITION p0 VALUES IN( (0,0), (NULL,NULL) ),
+	PARTITION p1 VALUES IN( (0,1), (0,2), (0,3), (1,1), (1,2) ),
+	PARTITION p2 VALUES IN( (1,0), (2,0), (2,1), (3,0), (3,1) ),
+	PARTITION p3 VALUES IN( (1,3), (2,2), (2,3), (3,2), (3,3) )
+);
+show tables;
```
Comment (fengttt) [inline test/cases/ddl/table_partition.test:87]: There need to have a show tables that prints out all columns, etc.   We need to at least print out some partition by information.

**Comment by qingxinhome** [comment]
> 

OK， I have added BVT test

**Review by iamlinjunhong — APPROVED** [review]

File: test/cases/ddl/table_partition.test (hunk 8 above)
Comment (qingxinhome) [reply test/cases/ddl/table_partition.test:85]: Yes, a metadata table needs to be added later to record partition information

File: pkg/sql/plan/base_binder.go [hunk 9]
```diff
@@ -150,8 +150,13 @@ func (b *baseBinder) baseBindExpr(astExpr tree.Expr, depth int32, isRoot bool) (
 			},
 		}, nil
 	case *tree.MaxValue:
-		err = errors.New("", fmt.Sprintf("expr max'%v' is not supported now", exprImpl))
-
+		return &Expr{
+			Expr: &plan.Expr_Max{
+				Max: &MaxValue{
+					Value: "maxvalue",
```
Comment (ouyuanning) [inline pkg/sql/plan/base_binder.go:156]: the string "maxvalue" is make any sence? may be empty is ok.

**Review by ouyuanning — APPROVED** [review]

**Review by aressu1985 — APPROVED** [review]

File: test/cases/ddl/table_partition.test (hunk 8 above)
Comment (fengttt) [reply test/cases/ddl/table_partition.test:85]: So you plan to do it later?   You must update issue, document there is unfinished work.

**Review by nnsgmsone — APPROVED** [review]

File: pkg/sql/plan/base_binder.go (hunk 9 above)
Comment (qingxinhome) [reply pkg/sql/plan/base_binder.go:156]: It is also possible to set it to null. I want to provide a station

File: test/cases/ddl/table_partition.test (hunk 8 above)
Comment (qingxinhome) [reply test/cases/ddl/table_partition.test:85]: OK，I will do it.

Comment (qingxinhome) [reply test/cases/ddl/table_partition.test:85]: I have created a subtask to record and handle this case，issue #4749

**Review by fengttt — APPROVED** [review]

**Pull request merged**